### PR TITLE
Add --since argument to docker logs cmd

### PR DESCRIPTION
**Pull request by ahmetb**
Added `--since` argument to `docker logs` command. Accept unix
timestamps and shows logs only created after the specified date.

Default value is 0 and passing default value or not specifying
the value in the request causes parameter to be ignored (behavior
prior to this change).

Updated:
- [x] cli: by default sends 0, otherwise unix timestamp
- [x] remote api: updated remote api and docs
- [x] engine: updated server to filter logs by date
- [x] integration-cli: added simple test and a test with -f and a future timestamp
- [x] docs: updated remote api, cli.md, manpage and completion scripts

Signed-off-by: Ahmet Alp Balkan

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks! I think this makes sense. One possible concern; currently jsonlogs is the only driver supported for `docker logs`, but there's a proposal for a rollover driver (https://github.com/docker/docker/pull/11485) that also can be used with `docker logs‘ and will introduce a `--tail` flag.

It can quickly become confusing from a UX which flag/option is supported by which driver. Perhaps someone has thoughts how to resolve that?

To summarize; I'm +1 on this, but a quick evaluation on the UX (for this **and** future drivers) may be wise.

**Comment by thaJeztah** [comment]
Oh, probably;

/cc @LK4D4

Because I think he's most involved with the logging drivers :)

**Comment by ahmetalpbalkan** [comment]
@thaJeztah yes I'm aware that the currently code is closely coupled to json all over the place. In my other PR #12226 I opened a way of enabling more logging drivers to be "read-able".

In an ideal world a logging driver would respond to read calls like "no I don't support reading" or "yes I support reading but not the --since parameter".

**Comment by thaJeztah** [comment]
Possibly `--since` can be implemented in the rollover driver as well, but that's another discussion.

Just wanted to mention the possible concern, apart from that :+1:

**Comment by vieux** [comment]
As we are adding `--since` why not add `--until` at the same time (like docker events)

**Comment by ahmetalpbalkan** [comment]
@vieux hmm maybe some other PR. I probably can get it done quickly after this. With this PR, I just wanted to see if this idea sounds ok.

**Comment by thaJeztah** [comment]
I'm +1 on a separate PR; smaller PRs = less discussion needed.

**Comment by ahmetalpbalkan** [comment]
Anyone?

**Comment by thaJeztah** [comment]
ping @LK4D4 :)

**Comment by icecrime** [comment]
Pushing to `2-needs-code-review`: @LK4D4 WDYT?

**Comment by ahmetalpbalkan** [comment]
(repushed to solve merge conflicts)

**Comment by runcom** [comment]
@ahmetalpbalkan needs to rebase again and errors about using old testing framework

**Comment by ahmetalpbalkan** [comment]
@runcom resolved merge conflicts again. Please review this folks. this is has been just sitting almost for a month now for no reason. 
cc:@LK4D4

File: api/client/utils.go [hunk 1]
```diff
@@ -343,3 +344,19 @@ func readBody(stream io.ReadCloser, statusCode int, err error) ([]byte, int, err
 	}
 	return body, statusCode, nil
 }
+
+// getTimestamp tries to parse given string as RFC3339 time
+// or Unix timestamp, if successful returns a Unix timestamp
+// as string otherwise returns value back.
+func getTimestamp(value string) string {
```
Comment (runcom) [inline api/client/utils.go]: why isn't this in pkg/timeutils? just wondering

File: api/client/utils.go [hunk 2]
```diff
@@ -343,3 +344,19 @@ func readBody(stream io.ReadCloser, statusCode int, err error) ([]byte, int, err
...
+	format := timeutils.RFC3339NanoFixed
+	loc := time.FixedZone(time.Now().Zone())
+	if len(value) < len(format) {
+		format = format[:len(value)]
+	}
+	if t, err := time.ParseInLocation(format, value, loc); err == nil {
+		return strconv.FormatInt(t.Unix(), 10)
+	} else {
```
Comment (runcom) [inline api/client/utils.go]: else here is redunant also

File: integration-cli/docker_cli_logs_test.go [hunk 3]
```diff
@@ -276,6 +277,78 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
...
+}
+
+func (s *DockerSuite) TestLogsSinceFutureFollow(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "-d", "busybox", "/bin/sh", "-c", `for i in $(seq 1 5); do date +%s; sleep 1; done`)
+	out, _, _, err := runCommandWithStdoutStderr(runCmd)
+	if err != nil {
+		c.Fatalf("run failed with errors: %s, %v", out, err)
+	}
+	cleanedContainerID := strings.TrimSpace(out)
+	defer deleteContainer(cleanedContainerID)
```
Comment (runcom) [inline integration-cli/docker_cli_logs_test.go]: this isn't needed anymore with gocheck

File: integration-cli/docker_cli_logs_test.go [hunk 4]
```diff
@@ -276,6 +277,78 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
...
 }
 
+func (s *DockerSuite) TestLogsSince(c *check.C) {
+	name := "testlogssince"
+	runCmd := exec.Command(dockerBinary, "run", fmt.Sprintf("--name=%s", name), "busybox", "/bin/sh", "-c", `date +%s; for i in $(seq 1 5); do sleep 1; echo log$i; done`)
+	out, _, _, err := runCommandWithStdoutStderr(runCmd)
+	if err != nil {
+		c.Fatalf("run failed with errors: %s, %v", out, err)
+	}
+	defer deleteContainer(name)
```
Comment (runcom) [inline integration-cli/docker_cli_logs_test.go]: remove this as well

**Comment by runcom** [comment]
code LGTM, also as already stated I'd like to have have an `until` flag too for consistency with other docker commands

File: integration-cli/docker_cli_logs_test.go [hunk 5]
```diff
@@ -276,6 +277,78 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
 	deleteContainer(cleanedContainerID)
 }
 
+func (s *DockerSuite) TestLogsSince(c *check.C) {
+	name := "testlogssince"
+	runCmd := exec.Command(dockerBinary, "run", fmt.Sprintf("--name=%s", name), "busybox", "/bin/sh", "-c", `date +%s; for i in $(seq 1 5); do sleep 1; echo log$i; done`)
+	out, _, _, err := runCommandWithStdoutStderr(runCmd)
```
Comment (runcom) [inline integration-cli/docker_cli_logs_test.go]: isn't `runCommandWithOutput` ok here since you don't care about stderr?

File: integration-cli/docker_cli_logs_test.go [hunk 6]
```diff
@@ -276,6 +277,78 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
...
+		c.Fatalf("run failed with errors: %s, %v", out, err)
+	}
+	defer deleteContainer(name)
+
+	outLines := strings.Split(out, "\n")
+	startUnix, _ := strconv.ParseInt(outLines[0], 10, 64)
+	since := startUnix + 3
+	logsCmd := exec.Command(dockerBinary, "logs", "-t", fmt.Sprintf("--since=%v", since), name)
+
+	out, _, _, err = runCommandWithStdoutStderr(logsCmd)
```
Comment (runcom) [inline integration-cli/docker_cli_logs_test.go]: `runCommandWithOutput`?

File: api/server/server.go [hunk 7]
```diff
@@ -594,9 +594,19 @@ func (s *Server) getContainersLogs(version version.Version, w http.ResponseWrite
 		return fmt.Errorf("Bad parameters: you must choose at least one stream")
 	}
 
+	var since time.Time
+	if r.Form.Get("since") != "" {
+		s, err := strconv.ParseInt(r.Form.Get("since"), 10, 64)
+		if err != nil {
+			return err
+		}
+		since = time.Unix(s, 0)
```
Comment (duglin) [inline api/server/server.go]: When I look at the other uses of "since" (for other commands) none of them call time.Unix(), but you do here.  Why?  I bet its probably correct, but I'm curious.

File: api/client/utils.go [hunk 8]
```diff
@@ -343,3 +344,19 @@ func readBody(stream io.ReadCloser, statusCode int, err error) ([]byte, int, err
...
+// getTimestamp tries to parse given string as RFC3339 time
+// or Unix timestamp, if successful returns a Unix timestamp
+// as string otherwise returns value back.
+func getTimestamp(value string) string {
+	format := timeutils.RFC3339NanoFixed
+	loc := time.FixedZone(time.Now().Zone())
+	if len(value) < len(format) {
+		format = format[:len(value)]
+	}
+	if t, err := time.ParseInLocation(format, value, loc); err == nil {
```
Comment (runcom) [inline api/client/utils.go]: nit but I'll just do
```
t, err := time.ParseInLocation(format, value, loc)
if err != nil {
    return value
}
return strconv.FormatInt(t.Unix(), 10)
```

File: daemon/logs.go [hunk 9]
```diff
@@ -101,9 +104,11 @@ func (daemon *Daemon) ContainerLogs(name string, config *ContainerLogsConfig) er
 				}
 				cLog = tmp
 			}
+
 			dec := json.NewDecoder(cLog)
 			l := &jsonlog.JSONLog{}
 			for {
+				l.Reset()
```
Comment (duglin) [inline daemon/logs.go]: Was this moved because sometimes the call to `dec.Decode(l)` didn't erase some fields?

Comment (runcom) [inline daemon/logs.go]: mmm I think because of the `continue` below

File: api/client/utils.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline api/client/utils.go]: yeah I guess we could move, this used to be buried in the code here, I just extracted into a method.

File: integration-cli/docker_cli_logs_test.go (hunk 5 above)
Comment (ahmetalpbalkan) [inline integration-cli/docker_cli_logs_test.go]: @runcom I guess true, I just copy pasted from the test above.

File: daemon/logs.go (hunk 9 above)
Comment (duglin) [inline daemon/logs.go]: Yea, but in looking at the code between here and there, nothing is set/changed in 'l' so I'm wondering why it needed to be moved.  I kind of wonder if this is fixing a potential bug that's been there for a while.

File: api/client/utils.go (hunk 8 above)
Comment (ahmetalpbalkan) [inline api/client/utils.go]: @runcom makes sense, I'll change. I extracted existing code into a method here. That's why I didn't touch it.

File: api/server/server.go (hunk 7 above)
Comment (ahmetalpbalkan) [inline api/server/server.go]: @duglin other usages in this file are either (1) converting to time.Time at some point later on or (2) the "since" is not a time.

I preferred converting and storing since to time.Time right here so that it doesn't get passed to stack below as int and has a special case s==0.

File: daemon/logs.go (hunk 9 above)
Comment (ahmetalpbalkan) [inline daemon/logs.go]: @duglin yes, it's moved because of the `continue`. This way we always reset the object when the loop start again. Otherwise I was gonna do `continue; l.Reset(); } ...` which is ugly. This doesn't affect the existing behavior.

**Comment by ahmetalpbalkan** [comment]
@runcom addressed comments.

**Comment by runcom** [comment]
code LGTM if janky's ok, I'm +1 on this and as already stated I think we should have `--until` as well like other docker commands

**Comment by ahmetalpbalkan** [comment]
Yeah I'll submit a PR for --until separately thanks.

**Comment by thaJeztah** [comment]
Needs one more LGTM before docs review (@duglin ?)

(Gonna be a bit insubordinate and add some notes in advance, otherwise I forget them :smile:)

File: docs/sources/reference/api/docker_remote_api_v1.19.md [hunk 10]
```diff
@@ -493,6 +493,7 @@ Query Parameters:
 -   **follow** – 1/True/true or 0/False/false, return stream. Default false
 -   **stdout** – 1/True/true or 0/False/false, show stdout log. Default false
 -   **stderr** – 1/True/true or 0/False/false, show stderr log. Default false
+-   **since** – UNIX epoch time to filter logs created since a timestamp. Default: 0.
```
Comment (thaJeztah) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: UNIX timestamp (integer) to filter logs. Specifying a timestamp will only output log-entries since that timestamp. Default: 0 (unfiltered).


(The UNIX epoch time is 01-01-1970; a UNIX *timestamp* is the number of seconds since that epoch)

File: docs/sources/reference/commandline/cli.md [hunk 11]
```diff
@@ -1616,6 +1616,7 @@ For example:
     Fetch the logs of a container
 
       -f, --follow=false        Follow log output
+      --since=""		Show logs since timestamp
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: "Show .. " doesn't align with the other lines here.

File: docs/sources/reference/commandline/cli.md [hunk 12]
```diff
@@ -1635,6 +1636,9 @@ timestamp, for example `2014-09-16T06:17:46.000000000Z`, to each
 log entry. To ensure that the timestamps for are aligned the
 nano-second part of the timestamp will be padded with zero when necessary.
 
+The `docker logs --since` argument shows logs of a container generated only
+since the specificed date. It can also be used with `--follow` and `--tail`.
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: s/specificed/specified/

date -> date/time or timestamp.

"It can also be used with" -> "The `--since` can be combined with the `--follow` and `--tail` options."

File: docs/sources/reference/api/docker_remote_api_v1.19.md (hunk 10 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: good point.

File: docs/sources/reference/commandline/cli.md (hunk 11 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/commandline/cli.md]: ah tabs vs spaces... fixed.

File: docs/sources/reference/commandline/cli.md (hunk 12 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/commandline/cli.md]: fixed. thx

File: docs/sources/reference/api/docker_remote_api_v1.19.md [hunk 13]
```diff
@@ -493,6 +493,8 @@ Query Parameters:
 -   **follow** – 1/True/true or 0/False/false, return stream. Default false
 -   **stdout** – 1/True/true or 0/False/false, show stdout log. Default false
 -   **stderr** – 1/True/true or 0/False/false, show stderr log. Default false
+-   **since** – UNIX timestamp (integer) to filter logs. Specifying a timestamp
```
Comment (duglin) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: this only takes the int form, not the other ones like `--since '2013-09-03'` ?

Comment (thaJeztah) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: Oh, good one, could be my bad? 😊

File: integration-cli/docker_cli_logs_test.go [hunk 14]
```diff
@@ -276,6 +277,76 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
 	deleteContainer(cleanedContainerID)
 }
 
+func (s *DockerSuite) TestLogsSince(c *check.C) {
+	name := "testlogssince"
+	runCmd := exec.Command(dockerBinary, "run", fmt.Sprintf("--name=%s", name), "busybox", "/bin/sh", "-c", `date +%s; for i in $(seq 1 5); do sleep 1; echo log$i; done`)
```
Comment (duglin) [inline integration-cli/docker_cli_logs_test.go]: For next time, you can make it: `"--name", name,` instead of using the Sprintf stuff.

File: integration-cli/docker_cli_logs_test.go [hunk 15]
```diff
@@ -276,6 +277,76 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
...
+	// Skip 2 seconds
+	unexpected := []string{"log1", "log2"}
+	for _, v := range unexpected {
+		if strings.Contains(out, v) {
+			c.Fatalf("unexpected log message returned=%v, since=%v\nout=%v", v, since, out)
+		}
+	}
+
+	// Test default
+	logsCmd = exec.Command(dockerBinary, "logs", "-t", "--since=0", name)
```
Comment (duglin) [inline integration-cli/docker_cli_logs_test.go]: Can you do me a favor and run this test twice, once with --since=0 and once w/o any --since at all just to make sure both return the same (log1..log5) results?

**Comment by duglin** [comment]
Just a couple of minor comments - other than that it looks good

File: integration-cli/docker_cli_logs_test.go (hunk 14 above)
Comment (duglin) [inline integration-cli/docker_cli_logs_test.go]: or even `"--name="+name`

File: docs/sources/reference/api/docker_remote_api_v1.19.md (hunk 13 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: @duglin remote API only takes Unix timestamps, not date formats. That's the case with since/until parameters in /events as well. I kept it consistent.

File: integration-cli/docker_cli_logs_test.go (hunk 15 above)
Comment (ahmetalpbalkan) [inline integration-cli/docker_cli_logs_test.go]: alrighty

File: integration-cli/docker_cli_logs_test.go (hunk 14 above)
Comment (ahmetalpbalkan) [inline integration-cli/docker_cli_logs_test.go]: sure

File: docs/sources/reference/api/docker_remote_api_v1.19.md (hunk 13 above)
Comment (duglin) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: are the docs for events wrong?
https://docs.docker.com/reference/commandline/cli/#examples_3

Comment (ahmetalpbalkan) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: @duglin those are CLI docs. CLI converts rfc3339 time format into Unix timestamp and sends to the remote API as unix ts. (We're commenting on the remote api doc file right now).

Comment (duglin) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: ok - sorry for the confusion

**Comment by duglin** [comment]
LGTM

**Comment by thaJeztah** [comment]
@runcom does this still LGTY? Because with two "LGTM"s, I think it can be moved to docs-review

**Comment by runcom** [comment]
@thaJeztah yep! I'm moving this to docs-review

File: docs/sources/reference/commandline/cli.md [hunk 16]
```diff
@@ -1635,6 +1636,10 @@ timestamp, for example `2014-09-16T06:17:46.000000000Z`, to each
 log entry. To ensure that the timestamps for are aligned the
 nano-second part of the timestamp will be padded with zero when necessary.
 
+The `docker logs --since` argument shows logs of a container generated only
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: Reading to again, perhaps make it more in line with the paragraphs above;

"The `docker logs --since` option will filter the output and only show log entries generated after the given date, specified as RFC 3339 or UNIX timestamp."

Would be nice to have some examples; can be similar to the "Show events in the past from a specified time" examples below https://docs.docker.com/reference/commandline/cli/#filtering

File: docs/sources/reference/commandline/cli.md [hunk 17]
```diff
@@ -1635,6 +1636,10 @@ timestamp, for example `2014-09-16T06:17:46.000000000Z`, to each
...
+since the specified date. The `--since` can be combined with the `--follow`
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: Oh, my bad; "The `--since` *option* ..."

Comment (ahmetalpbalkan) [inline docs/sources/reference/commandline/cli.md]: fixed

File: docs/sources/reference/commandline/cli.md (hunk 16 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/commandline/cli.md]: edited as you said, deferring examples to some next PR. :)

File: docs/sources/reference/commandline/cli.md [hunk 18]
```diff
@@ -1635,6 +1636,10 @@ timestamp, for example `2014-09-16T06:17:46.000000000Z`, to each
 log entry. To ensure that the timestamps for are aligned the
 nano-second part of the timestamp will be padded with zero when necessary.
 
+The `--since` option shows logs of a container generated only after
+the given date, specified as RFC 3339 or UNIX timestamp. The `--since` can
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: Looks like you missed this one; The `--since` *option*...

**Comment by thaJeztah** [comment]
Thanks, Ahmet, one comment, but other than that LGTM

cc @fredlf @jamtur01 @moxiegirl @SvenDowideit

File: docs/sources/reference/commandline/cli.md (hunk 18 above)
Comment (ahmetalpbalkan) [inline docs/sources/reference/commandline/cli.md]: ah, fixed.

**Comment by ahmetalpbalkan** [comment]
ping for attention

**Comment by thaJeztah** [comment]
ping @fredlf @jamtur01 @moxiegirl @SvenDowideit please review (Ahmet is trying to get back his "number one" contributor spot :wink:)

**Comment by jamtur01** [comment]
Docs LGTM

**Comment by thaJeztah** [comment]
(Thanks James)

Docs LGTM

**Comment by thaJeztah** [comment]
@ahmetalpbalkan merged \o/. Next up; `--until`?

**Comment by ahmetalpbalkan** [comment]
Thanks for reviewing folks.
@thaJeztah yep! or maybe #13107 first.
